pkg/pool-gyro/contracts/lib/GyroECLPMath.sol

```diff
@@ -0,0 +1,772 @@
+
+import { FixedPoint } from "@balancer-labs/v3-solidity-utils/contracts/math/FixedPoint.sol";
+
+import { SignedFixedPoint } from "./SignedFixedPoint.sol";
+import { GyroPoolMath } from "./GyroPoolMath.sol";
+
+/**
+ * @notice ECLP math library. Pretty much a direct translation of the python version.
+ * @dev We use *signed* values here because some of the intermediate results can be negative (e.g. coordinates of
+ * points in the untransformed circle, "prices" in the untransformed circle).
+ */
+library GyroECLPMath {
+    using SignedFixedPoint for int256;
+    using FixedPoint for uint256;
+    using SafeCast for uint256;
+    using SafeCast for int256;
+
+    error RotationVectorSWrong();
+    error RotationVectorCWrong();
+    error RotationVectorNotNormalized();
+    error AssetBoundsExceeded();
+    error DerivedTauAlphaNotNormalized();
+    error DerivedTauBetaNotNormalized();
+    error StretchingFactorWrong();
+    error DerivedUWrong();
+    error DerivedVWrong();
+    error DerivedWWrong();
+    error DerivedZWrong();
+    error InvariantDenominatorWrong();
+    error MaxAssetsExceeded();
+    error MaxInvariantExceeded();
+    error DerivedDsqWrong();
+
+    uint256 internal constant _ONEHALF = 0.5e18;
+    int256 internal constant _ONE = 1e18; // 18 decimal places
+    int256 internal constant _ONE_XP = 1e38; // 38 decimal places
+
+    // Anti-overflow limits: Params and DerivedParams (static, only needs to be checked on pool creation).
+    int256 internal constant _ROTATION_VECTOR_NORM_ACCURACY = 1e3; // 1e-15 in normal precision
+    int256 internal constant _MAX_STRETCH_FACTOR = 1e26; // 1e8   in normal precision
+    int256 internal constant _DERIVED_TAU_NORM_ACCURACY_XP = 1e23; // 1e-15 in extra precision
+    int256 internal constant _MAX_INV_INVARIANT_DENOMINATOR_XP = 1e43; // 1e5   in extra precision
+    int256 internal constant _DERIVED_DSQ_NORM_ACCURACY_XP = 1e23; // 1e-15 in extra precision
+
+    // Anti-overflow limits: Dynamic values (checked before operations that use them).
+    int256 internal constant _MAX_BALANCES = 1e34; // 1e16 in normal precision
+    int256 internal constant _MAX_INVARIANT = 3e37; // 3e19 in normal precision
+
+    // Invariant growth limit: non-proportional add cannot cause the invariant to increase by more than this ratio.
+    uint256 public constant MIN_INVARIANT_RATIO = 60e16; // 60%
+    // Invariant shrink limit: non-proportional remove cannot cause the invariant to decrease by less than this ratio.
+    uint256 public constant MAX_INVARIANT_RATIO = 500e16; // 500%
+
+    struct QParams {
+        int256 a;
+        int256 b;
+        int256 c;
+    }
+
+    /// @dev Enforces limits and approximate normalization of the rotation vector.
+    function validateParams(IGyroECLPPool.EclpParams memory params) internal pure {
+        require(params.s > 0 && params.s < _ONE, RotationVectorSWrong());
+        require(params.c > 0 && params.c < _ONE, RotationVectorCWrong());
```

Comment:
The original code is slightly different, and it admits both boundaries 0 and FP(1).

```suggestion
        require(0 <= params.s && params.s <= ONE, RotationVectorSWrong());
        require(0 <= params.c && params.c <= ONE, RotationVectorCWrong());
```